packages/core/contracts/oracle/implementation/SkinnyOptimisticOracle.sol

```diff
@@ -0,0 +1,658 @@
+     * @param _timestamp timestamp to identify the existing request.
+     * @param _ancillaryData ancillary data of the price being requested.
+     * @param _request price request parameters. The hash of these parameters must match with the request hash that is
+     * associated with the price request unique ID {requester, identifier, timestamp, ancillaryData}, or this method
+     * will revert.
+     * @return boolean indicating true if price exists and false if not.
+     */
+    function hasPrice(
+        address _requester,
+        bytes32 _identifier,
+        uint256 _timestamp,
+        bytes memory _ancillaryData,
+        Request memory _request
+    ) public override nonReentrant() returns (bool) {
+        bytes32 requestId = _getId(_requester, _identifier, _timestamp, _ancillaryData);
+        _validateRequestHash(requestId, _request);
+        OptimisticOracleInterface.State state =
+            _getState(_requester, _identifier, _timestamp, _ancillaryData, _request);
+        return
+            state == OptimisticOracleInterface.State.Settled ||
+            state == OptimisticOracleInterface.State.Resolved ||
+            state == OptimisticOracleInterface.State.Expired;
+    }
+
+    /**
+     * @notice Generates stamped ancillary data in the format that it would be used in the case of a price dispute.
+     * @param _ancillaryData ancillary data of the price being requested.
+     * @param _requester sender of the initial price request.
+     * @return the stamped ancillary bytes.
+     */
+    function stampAncillaryData(bytes memory _ancillaryData, address _requester)
+        public
+        pure
+        override
+        returns (bytes memory)
+    {
+        return _stampAncillaryData(_ancillaryData, _requester);
+    }
+
+    // Returns hash of unique request identifiers. This contract maps request ID hashes to hashes of the request's
+    // parameters.
+    function _getId(
+        address _requester,
+        bytes32 _identifier,
+        uint256 _timestamp,
+        bytes memory _ancillaryData
+    ) private pure returns (bytes32) {
+        return keccak256(abi.encode(_requester, _identifier, _timestamp, _ancillaryData));
+    }
+
+    // Returns hash of request parameters. These are mapped to the unique request ID to track a request's lifecycle.
+    function _getRequestHash(Request memory _request) private pure returns (bytes32) {
+        return
+            keccak256(
+                abi.encode(
+                    _request.proposer,
+                    _request.disputer,
+                    _request.currency,
+                    _request.settled,
+                    _request.proposedPrice,
+                    _request.resolvedPrice,
+                    _request.expirationTime,
+                    _request.reward,
+                    _request.finalFee,
+                    _request.bond,
+                    _request.customLiveness
+                )
```

Comment:
Not really a gas optimization, but would this be simpler?
```
return keccak256(abi.encode(_request));
```

Reply:
Yeah definitely more readable, fixed